include/vinecopulib/bicop/implementation/tll.ipp

```diff
@@ -251,14 +253,21 @@ TllBicop::fit(const Eigen::Matrix<double, Eigen::Dynamic, 2>& data,
   interp_grid_ = std::make_shared<InterpolationGrid>(grid_points, values);
 
   // compute effective degrees of freedom via interpolation ---------
-  Eigen::VectorXd infl_vec = ll_fit.col(1);
   // stabilize interpolation by restricting to plausible range
-  infl_vec = infl_vec.array().min(1.0).max(-1.0);
+  Eigen::VectorXd infl_vec = ll_fit.col(1).cwiseMin(1.3).cwiseMax(-0.2);
   Eigen::MatrixXd infl(m, m);
   infl = Eigen::Map<Eigen::MatrixXd>(infl_vec.data(), m, m).transpose();
   // don't normalize margins of the EDF! (norm_times = 0)
   auto infl_grid = InterpolationGrid(grid_points, infl, 0);
-  npars_ = infl_grid.interpolate(data).sum();
-  set_loglik(loglik(data, weights));
+  if ((var_types_[0] == "d") | (var_types_[1] == "d")) {
+      // for discrete, use mid ranks to compute EDF and log-likelihood
+      // (this is closer to "observations" than jittered or "upper" pseudo data)
+      psobs = 0.5 * (data.leftCols(2) + data.rightCols(2)).array();
+      npars_ = tools_eigen::unique(infl_grid.interpolate(psobs)).sum();
```

Comment:
Are we preventing the NaNs somewhow here?

Reply:
NaNs are removed [here](https://github.com/vinecopulib/vinecopulib/blob/discrete-no-dc/include/vinecopulib/bicop/implementation/class.ipp#L699).